### PR TITLE
Restore params from a checkpoint dir for ps instances

### PR DESCRIPTION
**Pull request by workingloong**
Implement #1507 & #1506

### CODE REVIEW

File: elasticdl/python/ps/parameter_server.py [hunk 1]
```diff
@@ -42,6 +46,26 @@ def __init__(self, args):
...
+
+        if not check_checkpoint_valid(checkpoint_dir_for_init):
+            raise ValueError("Invalid checkpoint directory")
+
+        self.parameters = restore_model_params_from_checkpoint(
+            checkpoint_dir_for_init, self.ps_id, self.num_ps_pods
+        )
+        self.parameters.init_status = True
+        self.logger.info(
+            "The version of restored parameters is %s"
```
Comment (brightcoder01) [inline elasticdl/python/ps/parameter_server.py]: parameters.version is a string type?

File: elasticdl/python/tests/pserver_servicer_test.py [hunk 2]
```diff
@@ -469,6 +469,63 @@ def test_save_parameters_to_checkpoint_file(self):
                 ["variables-0-of-1.chkpt"],
             )
 
+    def test_restore_parameters_from_checkpoint(self):
+        checkpoint_dir_for_init = (
+            "elasticdl/python/tests/testdata/functional_ckpt/version-100"
```
Comment (brightcoder01) [inline elasticdl/python/tests/pserver_servicer_test.py]: The test data directory for `functional_ckpt/version-100` is not included in this PR

**Review by terrytangyuan — COMMENTED** [review]
Can we reuse this for `Worker._update_local_model_params` when allreduce strategy is used?

File: docs/designs/sqlflow_integration.md [hunk 3]
```diff
@@ -65,7 +65,7 @@ PREDICT prediction_results_table
 WITH
     model.num_classes = 10,
     model.dataset_fn = "dataset_fn",
-    predict.checkpoint_filename_for_init = "modelv1.chkpt",
+    predict.checkpoint_dir_for_init = "v1/",
```
Comment (skydoorkai) [inline docs/designs/sqlflow_integration.md:68]: No need to add "/" in the string?

File: elasticdl/python/ps/parameter_server.py (hunk 1 above)
Comment (workingloong) [reply elasticdl/python/ps/parameter_server.py]: The parameters.version type is int.
https://github.com/sql-machine-learning/elasticdl/blob/c1bcb4562a285606523f5dc51672197f2986b740/elasticdl/python/ps/parameters.py#L38

The version in model message of grpc is also int.
https://github.com/sql-machine-learning/elasticdl/blob/c1bcb4562a285606523f5dc51672197f2986b740/elasticdl/proto/elasticdl.proto#L82-L86

File: docs/designs/sqlflow_integration.md (hunk 3 above)
Comment (workingloong) [reply docs/designs/sqlflow_integration.md:68]: Yes, "/" is not needed. I want to show "v1/" is a directory not a file by adding "/".

**Comment by workingloong** [comment]
> Can we reuse this for `Worker._update_local_model_params` when allreduce strategy is used?

Yes, the worker can save its owner parameters to a checkpoint after `Worker._update_local_model_params`. We can also create a parameters object for those parameters to save and restore checkpoint.

File: elasticdl/python/ps/parameter_server.py (hunk 1 above)
Comment (brightcoder01) [reply elasticdl/python/ps/parameter_server.py]: Yes, I'm wondering why we use "The version of restored parameters is %s"?

Comment (workingloong) [reply elasticdl/python/ps/parameter_server.py]: Oh, I will replace it with "%d"

File: elasticdl/python/common/save_utils.py [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+            model_pb = elasticdl_pb2.Model()
+            model_pb = load_pb_from_file(model_pb, shard_file_path)
+            if version is None:
+                version = model_pb.version
+            elif version != model_pb.version:
+                raise ValueError(
+                    "The versions in model shards are not consistency"
+                )
+
+        for embedding_info_pb in model_pb.embedding_table_info:
```
Comment (QiJune) [inline elasticdl/python/common/save_utils.py]: Put this inside the first for loop?

```python
for shard_file in variable_shard_files:
    ....
    for embedding_info_pb in model_pb.embedding_table_info:
        ....
```

File: elasticdl/python/common/save_utils.py [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+            shard_num: The total number of model shards, e.g. the total PS
+                instancecount using ParameterServerStrategy with multiple
+                PS instances.
+
+        Return:
+            parameters: A Parameter object which contains model version,
+                non-embedding parameters and embedding tables for the
+                PS instance with ps_id.
+        """
+        from elasticdl.python.ps.embedding_table import create_embedding_table
```
Comment (QiJune) [inline elasticdl/python/common/save_utils.py]: Please move this line to the top of this file.

File: elasticdl/python/common/save_utils.py (hunk 4 above)
Comment (workingloong) [reply elasticdl/python/common/save_utils.py]: Yes, Thanks! It is a mistake when I move the code.

File: elasticdl/python/common/save_utils.py [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+    return non_embedding_vars, embedding_table_values
+
+
+class Checkpoint(object):
+    def __init__(self, version, file):
+        self.version = version
+        self.file = file
+
+
+class CheckpointSaver(object):
```
Comment (terrytangyuan) [inline elasticdl/python/common/save_utils.py:68]: Can this implementation be generic instead of specific for multi-ps?

**Review by QiJune — APPROVED** [review]
Excellent!

File: elasticdl/python/common/save_utils.py (hunk 6 above)
Comment (workingloong) [reply elasticdl/python/common/save_utils.py:68]: The `CheckpointSaver` is not specific for multi-ps. It takes charge of saving the parameters including embedding and non-embedding variables using `proto.Model` in the current process. For all-reduce and other strategies, we can also use the implementation to save checkpoints. Using all-reduce strategy, any worker can create a `proto.Model` with its all parameters and call `CheckpointSaver.save` with `shard_index=0` and `shard_num=1`. We also can restore a `Parameters` instance from a checkpoint file and assign those parameters to the model in workers.

**Review by terrytangyuan — APPROVED** [review]

**Pull request merged**